### PR TITLE
Implemented new Difficulty Predictor model that uses ONNX Runtime instead of PMML to enable LDA in the model.

### PR DESCRIPTION
**Pull request by Brollyy**
#109 

The new model has way better predictive power on lower diffs and is slightly more consistent with 4-6 diffs.

I removed the "Treat maps as WIP" checkbox, since with the new metrics, the model always takes the time between first and last note, so there was no difference.

There were some issues with very hard maps, where the predicted difficulty was very low, so we've opted in to display a warning instead when the map parameters go outside of range on which the model was trained:
image

EDIT: after remodelling due to some implementation issues before, the new model is slightly improved. It uses these features:
1. Note density of the whole song.
2. Average time differences between notes.
3. Number of time differences below 2 seconds (essentially counts how many notes are "close" to each other to account for longer breaks).
4. 95th percentile of the local note densities in a rolling 4 second window (2 seconds before note, 2 seconds after note).
5. 40th percentile of time differences between notes.

### CODE REVIEW

File: Windows/DifficultyPredictorWindow.xaml.cs [hunk 1]
```diff
@@ -136,10 +138,10 @@ public partial class DifficultyPredictorWindow : Window {
         }
 
         private bool CheckParameterRanges(double noteDensity, double averageTimeDifference, int longestHighDensitySequence, double peakNoteDensity) {
-            return Helper.DoubleRangeCheck(noteDensity, 1.627074, 7.615101) &&
```
Comment (PKBeam) [inline Windows/DifficultyPredictorWindow.xaml.cs]: should probably put the magic numbers in the constants file

Comment (Brollyy) [reply Windows/DifficultyPredictorWindow.xaml.cs]: Done

**Review by PKBeam — APPROVED** [review]

File: Windows/DifficultyPredictorWindow.xaml.cs [hunk 2]
```diff
@@ -25,144 +26,123 @@ public partial class DifficultyPredictorWindow : Window {
...
-        */
-        private double GetNoteDensity(IEnumerable<Note> notes, double songDuration) {
-            return notes.Count() / songDuration;
+
+        private static double GetNoteDensity(IEnumerable<double> noteTimes, double songDuration) {
+            return noteTimes.Count() / songDuration;
         }
-        private List<double> GetLocalNoteDensity(IEnumerable<Note> notes, double songDuration, double globalBpm, double windowLength = 2.75, double step = 0.25) {
+
+        private static List<double> GetLocalNoteDensity(IEnumerable<double> timeSeries, double songDuration, double windowLength = 2.75, double step = 0.25) {
```
Comment (Nytilde) [inline Windows/DifficultyPredictorWindow.xaml.cs:76]: This is not equivalent to the python code

maybe something like

```csharp
private static List<double> GetNoteDensityPerWindow(List<double> timeArray, List<double> timeDiff, double windowLength = 2.75) {
    var cumulativeTime = timeDiff
        .Select((t, i) => timeDiff
        .Take(i + 1).Sum())
        .ToList();
    var startTimes = timeArray
        .Select(t => t - windowLength)
        .ToList();
    var startIndices = startTimes
        .Select(t => cumulativeTime.BinarySearch(t))
        .Select(idx => idx >= 0 ? idx : ~idx)
        .ToList();
    var endIndices = timeArray
        .Select(t => cumulativeTime
        .BinarySearch(t))
        .Select(idx => idx >= 0 ? idx : ~idx)
        .ToList();
    return startIndices
        .Zip(endIndices, (start, end) => (end - start) / windowLength)
        .ToList();
}
```

File: Windows/DifficultyPredictorWindow.xaml.cs [hunk 3]
```diff
@@ -25,144 +26,123 @@ public partial class DifficultyPredictorWindow : Window {
...
-                        // higher is better
-                        variety.Add(-1 * score);
 
-                    }
+        private static int GetLongestHighDensitySequence(List<double> timeDiffs) {
+            var uniqueValues = timeDiffs.Distinct().ToList();
+            var indices = uniqueValues.Select(val => timeDiffs.IndexOf(val)).ToList();
+            var counts = uniqueValues.Select(val => timeDiffs.Count(t => t == val)).ToList();
+
+            var matrix = new int[uniqueValues.Count, timeDiffs.Count];
```
Comment (Nytilde) [inline Windows/DifficultyPredictorWindow.xaml.cs:100]: maybe
```csharp
var matrix = uniqueValues.Select((val, i) => {
    var idx = indices[i];
    var count = counts[i];
    return Enumerable.Range(0, timeDiffs.Count)
        .Select(j => j >= idx && j < idx + count ? 1 : 0)
        .ToArray();
}).ToArray();
```

File: Windows/DifficultyPredictorWindow.xaml.cs [hunk 4]
```diff
@@ -25,144 +26,123 @@ public partial class DifficultyPredictorWindow : Window {
...
+
+                for (int j = idx; j < idx + count; j++) {
+                    matrix[valIndex, j] = 1;
                 }
-                windowLower += step;
-                windowUpper += step;
             }
-            return Helper.GetQuantile(variety, 0.75);
+
+            var rowSums = Enumerable
```
Comment (Nytilde) [inline Windows/DifficultyPredictorWindow.xaml.cs:112]: with
```csharp
var rowSums = matrix.Select(row => row.Sum()).ToList();
```

File: Windows/DifficultyPredictorWindow.xaml.cs [hunk 5]
```diff
@@ -25,144 +26,123 @@ public partial class DifficultyPredictorWindow : Window {
...
+                .Range(0, uniqueValues.Count)
+                .Select(i => Enumerable.Range(0, timeDiffs.Count).Sum(j => matrix[i, j]))
+                .ToList();
+
+            var maxDensityIndex = rowSums.IndexOf(rowSums.Max());
+            var maxDensity = (double)rowSums[maxDensityIndex] / timeDiffs.Count;
+
+            var seriesStart = Array.IndexOf(matrix.Cast<int>().ToArray(), 1, maxDensityIndex * timeDiffs.Count);
+            var seriesEnd = Array.LastIndexOf(matrix.Cast<int>().ToArray(), 1, (maxDensityIndex + 1) * timeDiffs.Count - 1);
```
Comment (Nytilde) [inline Windows/DifficultyPredictorWindow.xaml.cs:121]:
```csharp
var seriesStart = Array.IndexOf(matrix[maxDensityIndex], 1);
var seriesEnd = Array.LastIndexOf(matrix[maxDensityIndex], 1);
```
should be cheaper

File: Windows/DifficultyPredictorWindow.xaml.cs [hunk 6]
```diff
@@ -25,144 +26,124 @@ public partial class DifficultyPredictorWindow : Window {
...
-                        // L1-normalise or normalise for the amount of notes
-                        var normLocalVariety = Helper.LpNormalise(localVariety, 1);
-                        // maps with higher column variety will have a distribution closer to [.25, .25, .25, .25]
-                        var score = Helper.LpDistance(normLocalVariety, new List<double>() { 0.25, 0.25, 0.25, 0.25 }, 2);
-                        // higher is better
-                        variety.Add(-1 * score);
-
-                    }
+
+        private static int GetLongestHighDensitySequence(List<double> timeDiffs) {
```
Comment (Brollyy) [inline Windows/DifficultyPredictorWindow.xaml.cs:96]: As discussed offline, we need to reimplement this function and retrain the model - apart from a bug in the Python code where `series_end` calculates an index of flipped array and doesn't flip it back, what this function calculates is essentially just how often mode value occurs in `timeDiffs`, e.g.:
```
time_diffs_no_zeros=[1, 2, 1, 3, 4, 1, 4]

unique_values=array([1, 2, 3, 4]) indices=array([0, 1, 3, 4], dtype=int64) counts=array([3, 1, 1, 2], dtype=int64)
matrix=array([[1, 1, 1, 0, 0, 0, 0],
       [0, 1, 0, 0, 0, 0, 0],
       [0, 0, 0, 1, 0, 0, 0],
       [0, 0, 0, 0, 1, 1, 0]])
row_sums=array([3, 1, 1, 2]) max_density_index=0
series_start=0 series_end=2
len(get_longest_low_value_sequence(time_diffs_no_zeros)) = 3
```
but the intention was to quantify what's the longest "burst" section in the map.

**Review by PKBeam — APPROVED** [review]

**Review by Nytilde — APPROVED** [review]

**Pull request merged**